### ISSUE TITLE
Misuse of Suggests

### ISSUE THREAD

**Issue by eddelbuettel** [r-dbi/RSQLite#202]
The package uses Suggests: when it either means Depends (for lack of conditioning) or, rather, fails to tests for presence of a package only suggested:

```
* using log directory ‘/tmp/RcppDepends/RSQLite.Rcheck’
* using R version 3.3.3 (2017-03-06)
* using platform: x86_64-pc-linux-gnu (64-bit)
* using session charset: UTF-8
* using options ‘--no-manual --no-vignettes’
* checking for file ‘RSQLite/DESCRIPTION’ ... OK
* this is package ‘RSQLite’ version ‘1.1-2’
* package encoding: UTF-8
* checking package namespace information ... OK
* checking package dependencies ... NOTE
Package suggested but not available for checking: ‘DBItest’
* checking if this is a source package ... OK
[....]
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ... ERROR
  Running ‘testthat.R’
Running the tests in ‘tests/testthat.R’ failed.
Complete output:
  > library(testthat)
  > library(RSQLite)
  > 
  > test_check("RSQLite")
  Error in loadNamespace(name) : there is no package called 'DBItest'
  Calls: test_check ... tryCatch -> tryCatchList -> tryCatchOne -> <Anonymous>
  Execution halted
* checking for unstated dependencies in vignettes ... OK
```

You *cannot* assume that a suggested is present.

**Comment by krlmlr** [comment]
Thanks. I haven't found any hint in R-exts that we should check presence of suggested packages in the tests.

> The ‘Suggests’ field uses the same syntax as ‘Depends’ and lists packages that are not necessarily needed. This includes packages used only in examples, tests or vignettes (see Writing package vignettes), and packages loaded in the body of functions. E.g., suppose an example from package foo uses a dataset from package bar. Then it is not necessary to have bar use foo unless one wants to execute all the examples/tests/vignettes...

This kind of implies that if one wants to execute all tests then it's necessary to install the suggested package. What am I missing?

**Comment by eddelbuettel** [comment]
Keep reading.  A few paras below:

> The general rules are
> 
> - A package should be listed in only one of these fields.
> - Packages whose namespace only is needed to load the package using library(pkgname) should be listed in the ‘Imports’ field and not in the ‘Depends’ field. Packages listed in imports or importFrom directives in the NAMESPACE file should almost always be in ‘Imports’ and not ‘Depends’.
> - Packages that need to be attached to successfully load the package using library(pkgname) must be listed in the ‘Depends’ field.
> - **All packages that are needed] to successfully run R CMD check on the package must be listed in one of ‘Depends’ or ‘Suggests’ or ‘Imports’. Packages used to run examples or tests conditionally (e.g. via if(require(pkgname))) should be listed in ‘Suggests’ or ‘Enhances’. (This allows checkers to ensure that all the packages needed for a complete check are installed.)**

You didn't use that package conditionally.  So as I said: either make it a Depends, or condition on it.

**Comment by krlmlr** [comment]
Agreed that I'm unconditionally using DBItest. The last bullet point, the one that you highlighted, seems to address only packages used to run tests conditionally, but I'm not doing that. I have listed DBItest in "Suggests"; I agree that I could also list it in "Enhances" if I decided to make the test conditional, but I'd rather avoid that.

**Comment by eddelbuettel** [comment]
I asked you two times already to either
- move it to Depends
- or prefix its use with `if (requireNamespace())`
What is so difficult about the latter?

I am tired of chasing packages like this and will just blacklist them from my revdeps.

**Comment by krlmlr** [comment]
We could have had a much nicer discussion if you had included your motivation up front.

Anyway, DBItest is a very lightweight dependency that is unlikely to pull in any more stuff you don't have to install anyway for the imports of the other packages. Would you please consider just installing the suggests (also for your other revdeps), even if it takes a gig on your drive and an extra hour of machine time? This is what I do when I revdep-check RSQLite itself, or dplyr..

**Comment by eddelbuettel** [comment]
Actually 1.3.1.3 is even stronger:

> Note that someone wanting to run the examples/tests/vignettes may not have a suggested package available (and it may not even be possible to install it for that platform). **The recommendation used to be to make their use conditional via if(require("pkgname"))): this is fine if that conditioning is done in examples/tests/vignettes.**

But I argued unsuccessfully for this before with other people;  to me it remains a bug and I will now avoid these packages.   Life being short and all that.

**Comment by krlmlr** [comment]
1.1.3.1 makes a case, but the DBItest package is neither difficult to obtain nor difficult to install.

The DBItest and testthat packages are an integral part of the tests in that package, it's next to pointless to run the tests without them. Making package use conditional weakens the test. If you don't want to run the tests, don't run the tests (or don't check the package), but this actually weakens your revdep check. What's the use of the revdep check if it succeeds on your machine but fails on CRAN? Adding these packages to Depends or Imports would be an even stronger packaging smell.

Sorry, but I don't see a good solution here other than failing the test if the package is missing, and that's the way it's currently implemented.

A final note: To deal with failing packages, I'm usually doing two revdep checks, one with the CRAN version, and the other with the release candidate, and then compare the outputs of R CMD check. Works well enough for me, and would most likely also work for this particular package. [diff.sh](https://github.com/hadley/dplyr/blob/d735087bff45955d98e05bcec4eff978ff3f3ca9/revdep/diff.sh)

**Comment by eddelbuettel** [comment]
> it's next to pointless to run the tests without them.

You just explained to yourself that it is really a Depends.  

I am done with:  http://dirk.eddelbuettel.com/blog/2017/03/22#suggests_is_not_depends

**Comment by eddelbuettel** [comment]
And it is astonishing how you continue to refuse to acknowledge that the language has a perfectly fine mechanism for this:  `if (requireNamespace(...))`.

**Comment by gaborcsardi** [comment]
I think the main reason of the frustration is that `Suggests` is just **really** messed up. It simply cannot capture the roles of the various dependencies.

I agree with Dirk, that according to the CRAN/R-core rules, if `DBItest ` is in `Suggests`, then the package should check for it in the code using `if (requireNamespace(...))`.

But in my opinion this is not a good solution if the package is suggested for use in the tests. If you have a `requireNamespace` safeguard on the tests (most often this would be for `testthat` or `RUnit`), and the `Suggest`ed package is not available, then the tests are just skipped, which is rarely what you want. When you are testing a package, the last thing you want to do is skipping the test cases, right?

IMO the best solution would be to deprecate `Suggests` completely, and introduce two new dependency types:
* `OptionalImports`: this is for deps, that are optionally needed for the *user* of the package. E.g. `igraph` can plot networks using `tcltk`, but this is an optional feature, and <1% of the package, so this would be in `OptionalImports`.
* `DevImports`: this is for deps that are needed for developing the package. E.g. `knitr` to build a vignette, or the package that provides the vignette style files. `testthat` and `RUnit` would belong here, too.

[I was also thinking about having `BuildImports`, for packages that are needed for building the package, but now I think that these could simply go into `DevImports`. Building the package is often part of developing the package in general.]

Having new dependency types sounds like a pain, but I think forbidding `Suggests` for future packages, and having these two added, will actually simplify the choice between the dependencies, because it just describes the intent better. (Especially if we could come up with a bit better names than `OptionalImports` and `DevImports`.

cc @jimhester, @hadley in case they are interested

**Comment by krlmlr** [comment]
Thanks @gaborcsardi. Totally agree with the first paragraph. The third paragraph describes exactly why I'm reluctant to follow the rules described in the second paragraph.

What is the role of `Enhances` in your scenario? Can a package be listed in both `OptionalImports` and `DevImports`?

**Comment by gaborcsardi** [comment]
> What is the role of Enhances in your scenario? 

I am not sure. I don't really understand why `Enhances` is needed currently to be honest. AFAICT we can deprecate `Enhances` as well, in favour of `OptionalImports`, but I am not entirely sure.

> Can a package be listed in both OptionalImports and DevImports?

Sure, it can be listed in both. (Although it won't happen too much, probably.)

**Comment by jimhester** [comment]
I agree with @gaborcsardi, `Suggests` as currently used has two separate uses, which is the root cause of this issue.

I think `Enhances` is currently useless and should just be deprecated completely, it only serves to confuse the situation.

**Comment by eddelbuettel** [comment]
`Enhances` must die.  No purpose.  Was once meant for the same as `Recommends` (ie an actual recommendation of "could be useful along with") but flowing the other way. We have it in Debian too, and it is similarly underused.   "In theory" useful for discovering package relationships.  In practice, well, ...

I agree with the need for new semantics.  But I neither sure about the suggested names, nor do I think Suggests must be deprecated.  And lastly, I feel we all *utterly waste precious time* because without CRAN and R Core nothing will change. 

Where I started this ticket asking one maintainer for a one-minute change.  Alas ...

**Comment by gaborcsardi** [comment]
> nor do I think Suggests must be deprecated

What would you use it for?

> And lastly, I feel we all utterly waste precious time because without CRAN and R Core nothing will change.

I agree it is a very long shot. But if multiple people keep saying the same, maybe something will happen eventually.

**Comment by eddelbuettel** [comment]
What it is currently used for:  optional use.

**Comment by jangorecki** [comment]
I had the same problem as @eddelbuettel. @gaborcsardi I proposed new field to DESCRIPTION in https://stat.ethz.ch/pipermail/r-devel/2016-April/072526.html but no actions were taken. There is no way to really automate and scale revdeps check without addressing this issue.

**Comment by gaborcsardi** [comment]
> There is no way to really automate and scale revdeps check without addressing this issue.

There is, just install all suggested packages.

**Comment by jangorecki** [comment]
@gaborcsardi it is a way to automate but not scale

**Comment by gaborcsardi** [comment]
Then you can just do `R CMD check --no-tests`, and you don't need to install the suggested packages.

**Comment by gaborcsardi** [comment]
To be more constructive, how about we introduce a `SuggestsNote` field, in which we can specify which of the suggested packages are development dependencies? E.g.

```
SuggestsNote: DevDependencies: testthat, mockery
```

Then you would need to install the suggested packages that are dev dependencies. The other suggested packages are for optional use, and you don't need to install them.

Would this help? We can start using this without CRAN's approval, I believe.

**Comment by eddelbuettel** [comment]
This thread is boring as we continue to talk past each other, and useless because we cannot change packaging standards _on the CRAN side_ without CRAN.

I opened a ticket because I saw a one-minute change a maintainer could have made without effect on his package.  That maintainer disagreed.  Let's move one.

For what it is worth packages by more than half the participants in this thread behave correctly.  A few others do not.  Instead of fighting windmills, why don't we move on?   I updated by testing script and now blacklist packages _that fail to work under a specification which eg @jangorecki and I see as preferable_.  There is no right or wrong here -- even CRAN agrees that Suggests _should not_ be required but installs them anyway.  To each their own -- I won't for _my tests_ as I am free to set my rules there.

Peace out, and un-subcribing again from the thread.
